documentation/dataplane/dash-routing-actions.md

```diff
@@ -0,0 +1,197 @@
+# DASH routing actions
+
+As [DASH packet flow HLD](../general/dash-sai-pipeline-packet-flow.md) describes, routing actions are the fundamental building blocks that describe the packet transformations and form the high level scenarios in DASH. And this doc is used to describe the routing action definitions in details.
+
+1. [Overview](#overview)
+2. [Stage transition actions](#stage-transition-actions)
+   1. [Basic stage transition actions](#basic-stage-transition-actions)
+      1. [`drop` action](#drop-action)
+      2. [`trap` action](#trap-action)
+   2. [Matching stage transition actions](#matching-stage-transition-actions)
+      1. [IP-based stage transition actions](#ip-based-stage-transition-actions)
+         1. [`lpmrouting` action](#lpmrouting-action)
+         2. [`maprouting` action](#maprouting-action)
+      2. [Port-based stage transition actions](#port-based-stage-transition-actions)
+         1. [`portmaprouting` action](#portmaprouting-action)
+3. [Packet transformation actions](#packet-transformation-actions)
+   1. [`staticencap` action](#staticencap-action)
+   2. [`tunnel` action](#tunnel-action)
+   3. [`4to6` action](#4to6-action)
+   4. [`6to4` action](#6to4-action)
+   5. [`nat` action](#nat-action)
+
+## Overview
+
+On high level, there are 2 types of routing actions:
+
+- Packet transformation actions:
+  - These actions are used to transform the packet header fields, such as modify the packet fields, adding encaps, etc.
+  - These actions can only be used in the routing types that specified in the `routing_type` field in each match stage entry.
+- Stage transition actions:
+  - These actions are used to describe the stage transitions in the DASH pipeline, such as transit to maprouting stage.
+  - This actions can only be used in the routing types that specified in the `transition` field in each match stage entry.
+
+For how routing action works, please refer to the ["Routing action and routing types" section in DASH packet flow HLD](../general/dash-sai-pipeline-packet-flow.md#57-routing-actions-and-routing-types).
+
+## Stage transition actions
+
+### Basic stage transition actions
+
+#### `drop` action
+
+- Actions:
+  - Drop the packet after the entry is matched.
+
+#### `trap` action
+
+- Actions:
+  - Trap the packet to CPU after the entry is matched.
+
+### Matching stage transition actions
+
+All matching stage transition actions will support the following parameters:
+
+- `default_routing_type`: If no entry is found, use this routing type to route the packet. If `default_routing_type` is not set, the packet will be dropped by default.
+- `stage_index`: If there are multiple stage of the same stage type, use this field to specify which stage to transit to. The index is starting from 0 and by default 0.
```

Comment:
Can you please elaborate on multiple stage requirements a bit more. Today, the pipeline has a single routing and mapping stage along with multiple ACL stages. If there can be multiple routing stages, how is the action resolution expected to happen? Would there be a priority attached to each stage or the later stage would override the conflicting action from an earlier stage?

Reply:
It is actually similar to what you are thinking, and the process is very simple. The details are captured in the "Action Publishing" and "Metadata Publishing" sections, which covers the action itself and action parameters.

Say, we have 2 stage of LPM stage, and the process will be mostly like this:

- First, we create the LPM entries in the 2 LPM stages:
  - The stage index can be set by an attribute. Essentially, we can think it is LPM0 and LPM1.
  - Entry0 in LPM0 will have transition set to `lpmrouting` with `stage_index` set to 1, then when it matches, the second LPM stage will be invoked.
  - Entry1 in LPM1 can have the `routing_type` set, just like today and points to a list of actions, say: `staticencap`.
- Then, when a packet shows up, it matches Entry0, 
  - Transition info will be populated to metadata and make the pipeline skipping the internal stages (if any), and execute the second LPM.
  - Other metadata will also be populated, say, `underlay_dip` and etc.
- Then, the packet matches Entry0 in LPM1,
  - Transition info is not set, so we are reaching the end of matching stages.
  - Routing type is set, and `staticencap` action will be populated into metadata.
- When reaching ActionApply stage, the staticencap action will be executed, using the `underlay_dip` from the metadata bus.

More details and pseudo code can be found in "Stage skipping" and "Multi-stage chainning" sections.